File: imagelab-frontend/src/types/pipeline.ts
export interface PipelineStep {
  type: string;
  params: Record<string, unknown>;
  id?: string;
}

export interface PipelineRequest {
  image: string;
  image_format: string;
  pipeline: PipelineStep[];
  debug?: boolean;
}

export interface StepTiming {
  step: number;
  operator_type: string;
  duration_ms: number;
}

export interface PipelineTimings {
  total_ms: number;
  steps: StepTiming[];
}

export interface DebugStepState {
  step: number;
  block_id: string | null;
  operator_type: string;
  image: string;
  duration_ms: number;
}

export interface PipelineResponse {
  success: boolean;
  image?: string;
  image_format?: string;
  error?: string;
  step?: number;
  timings?: PipelineTimings;
  debug_states?: DebugStepState[];
}

File: imagelab-backend/app/models/pipeline.py
from pydantic import BaseModel


class PipelineStep(BaseModel):
    type: str
    params: dict = {}
    id: str | None = None


class PipelineRequest(BaseModel):
    image: str
    image_format: str = "png"
    pipeline: list[PipelineStep]
    debug: bool = False


class StepTiming(BaseModel):
    step: int
    operator_type: str
    duration_ms: float


class PipelineTimings(BaseModel):
    total_ms: float
    steps: list[StepTiming]


class DebugStepState(BaseModel):
    """Snapshot of the image after a single operator executes."""

    step: int
    block_id: str | None = None
    operator_type: str
    image: str
    duration_ms: float


class PipelineResponse(BaseModel):
    success: bool
    image: str | None = None
    image_format: str | None = None
    error: str | None = None
    step: int | None = None
    timings: PipelineTimings | None = None
    debug_states: list[DebugStepState] | None = None
